### ISSUE TITLE
PCE 6 buttons pad issue

### ISSUE THREAD

**Issue by sk8er000** [bootsector/LLOAD#9]
I've made 2 cable adapters for PC-Engine following the schemes. 
If I use a regular 2 buttons pad they work with no issues but if I try to use a 6 buttons pad (I've tested the one that came with the PC-Duo RX) it detects the UP and LEFT directions always pressed.
If I use the pad on the PC-Duo RX, PCE or Supergrafx it works

**Comment by bootsector** [comment]
Could you please test with the attached firmware and let me know the results?
[LLAMA.hex.zip](https://github.com/bootsector/LLAMA/files/4917843/LLAMA.hex.zip)

**Comment by sk8er000** [comment]
Is there a suggested update procedure?

**Comment by bootsector** [comment]
Are you using the DIY version with the Arduino Leonardo?

**Comment by sk8er000** [comment]
I'm using the version assembled by Bruno Freitas

**Comment by bootsector** [comment]
That would be me! :)

What operating system are you using so I can give you the right steps?

**Comment by sk8er000** [comment]
Didn't know that :D
I mainly use Windows 10 but I can use other OS if needed

**Comment by bootsector** [comment]
1 - Install the DFU driver by right clicking the dpinst64.exe installer in the drivers directory of the attached zip and running as admin  (assuming you have a 64bit Windows OS).

2 - Connect the P1 DFU jumper and connect the P1 USB to your PC via USB

3 - Double click the flash.bat file for updating the firmware.

4 - Repeat 2 and 3 for updating the second port 

Let me know how it goes!

[flash.zip](https://github.com/bootsector/LLAMA/files/4918518/flash.zip)

**Comment by sk8er000** [comment]
Thank you for the guide.
I've flashed the new fw!
Now the up direction works fine but the left direction is still pressed. If I press the right direction on the test pad ROM the left direction is not pressed anymore but neither the right direction.
The 2 buttons pad is still working with no issues

**Comment by bootsector** [comment]
Aside of the left direction being pressed, all the other buttons work correctly?

**Comment by sk8er000** [comment]
Yes they're working fine but I've discovered a strange behaviour:
-Actually I have the P1 with new fw and P2 with the old one
- 2 buttons controller works on both
- 6 buttons controller on older fw detects up and left always pressed on new one only left (all other signal are working, up direction is working only on the new fw)
- if I loose very much (untill they're barely connected) the connection between the adaptor and the pad on both firmwares everything is working.

Tested with 2 different pads and 2 different adapters

**Comment by bootsector** [comment]
Hmm... That's a really odd behavior. It's leaning to a bad connection somewhere. Only change from original fw to the new one I've sent you for testing is decreasing the frequency we talk to the controller. I'm attaching here a new hex so you can flash and test (now with an even slower frequency)
[LLAMA.hex.zip](https://github.com/bootsector/LLAMA/files/4918812/LLAMA.hex.zip)

**Comment by sk8er000** [comment]
It's still working in the same way.
The strange thing is that there are no issues with the regular pad on the directions and that the UP direction was fixed by the firmware. Otherwise the issue was obviously related to the wiring.
Since on the github there is only the scheme for the TG16 port adapter I've used a second scheme to traspose the TG16 signal to the PCE ones:
![pce](https://user-images.githubusercontent.com/35069659/87427553-3eca9a00-c5e1-11ea-88c2-f48e33621942.jpg)

the other strange thing if it's a connection problem is the same on both adapter and using 2 different pads (same model)

**Comment by bootsector** [comment]
Could you please give it a try with this one and let me know? 
[LLAMA.hex.zip](https://github.com/bootsector/LLAMA/files/4918953/LLAMA.hex.zip)

**Comment by sk8er000** [comment]
Now both models (2 and 6 buttons) are working with no issues!

**Comment by sk8er000** [comment]
Edit
Sometimes the left button is detected even if not pressed but is mainly working

**Comment by bootsector** [comment]
Just pushed this commig: https://github.com/bootsector/LLAMA/commit/6a01088b381b4b94614d69b81a3c75c0e6b575ef

I think there's something odd with the pullup resistor connected to your 6 button controller's left button.

**Comment by bootsector** [comment]
This is the latest commit but I've increased the delay from 20us to 40us. Could you please check if left button is still flaky?
[LLAMA.hex.zip](https://github.com/bootsector/LLAMA/files/4919028/LLAMA.hex.zip)

**Comment by sk8er000** [comment]
It seems to work, I'm sorry for all the trouble, unfortunately I've got only 2 6 buttons pad that were bundled with my pc duo Rx and I don't have other 6 buttons pad to test. This models are considered to have a clunky dpad, I'll try to open and inspect them but now it's much better than before! Thank you

**Comment by bootsector** [comment]
No problem. Let's keep this issue opened until we get this 100% figured out. Does it work better with latest zip or same thing as the previous one?

**Comment by sk8er000** [comment]
Now it's working a lot better sometimes if I hold firmly right it's been detected like if it's an autofire button (it's repeatedly pressed very fast instead of being firmly pressed). But it seems a pad issue rather than a fw issue (I'll check it soon).
Tested other 4 original pce 2 buttons pad (various models) and everyone is working properly.

**Comment by bootsector** [comment]
I think that it could be a good idea to check if the VCC on your DIN/mini-DIN connectors are really giving out 5V too.
